common/types/offer.go

```diff
@@ -42,30 +44,53 @@ func NewOffer(
 	exRate *coins.ExchangeRate,
 	ethAsset EthAsset,
 ) *Offer {
-	var buf [16]byte
-	if _, err := rand.Read(buf[:]); err != nil {
+	var n [8]byte
+	if _, err := rand.Read(n[:]); err != nil {
 		panic(err)
 	}
-	return &Offer{
+
+	offer := &Offer{
 		Version:      *CurOfferVersion,
-		ID:           sha3.Sum256(buf[:]),
 		Provides:     coin,
 		MinAmount:    minAmount,
 		MaxAmount:    maxAmount,
 		ExchangeRate: exRate,
 		EthAsset:     ethAsset,
+		Nonce:        binary.BigEndian.Uint64(n[:]),
+	}
+
+	offer.setID()
+	return offer
+}
+
+func (o *Offer) setID() {
+	if !IsHashZero(o.ID) {
+		panic("offer ID is already set")
 	}
+
+	o.ID = o.hash()
+}
+
+func (o *Offer) hash() Hash {
+	b := append([]byte(o.Version.String()), []byte(o.Provides)...)
+	b = append(b, []byte(o.MinAmount.Text('f'))...)
+	b = append(b, []byte(o.MaxAmount.Text('f'))...)
+	b = append(b, []byte(o.ExchangeRate.String())...)
+	b = append(b, o.EthAsset[:]...)
+	b = append(b, []byte(fmt.Sprintf("%d", o.Nonce))...)
+	return sha3.Sum256(b)
```

Comment:
Should we have comma separators to make it a human readable string?

Reply:
I don't think we need it to be human readable. I wanted to have the minimum amount of bytes in here as possible, so there's less that can be manipulated

Reply:
The commas, assuming no field can actually contain a comma, actually make it harder to manipulate. Without the commas, a min value of 111 and a max value of 222 becomes the same hash as a min value of 1 with a max value of 11222.